drone-webdriver/src/main/java/org/jboss/arquillian/drone/webdriver/binary/downloading/source/XmlStorageSource.java

```diff
@@ -0,0 +1,217 @@
+            log.warning("Date " + date + " of content " + storageUrl + key
+                            + " could not have been parsed. This content will be omitted. See the exception msg: "
+                            + e.getMessage());
+        }
+        return null;
+    }
+
+    protected String getContentOfFirstElement(Element element, String tagName) {
+        NodeList elementsByTagName = element.getElementsByTagName(tagName);
+        if (elementsByTagName.getLength() == 0) {
+            return "";
+        }
+        return elementsByTagName.item(0).getTextContent();
+    }
+
+    @Override
+    public ExternalBinary getReleaseForVersion(String requiredVersion) throws Exception {
+        this.driverEntries = retrieveAllDriversEntries();
+        final List<DriverEntry> matched = driverEntries
+            .stream()
+            .filter(driverEntry -> {
+                final String expectedKeyRegex = getExpectedKeyRegex(requiredVersion, driverEntry.getLocation());
+                final boolean matches = driverEntry.getKey().matches(expectedKeyRegex);
+                return matches;
+            })
+            .collect(Collectors.toList());
+
+        if (matched.size() == 0) {
+            throw new IllegalStateException(
+                "There wasn't found any binary with the key matching regex "
+                    + getExpectedKeyRegex(requiredVersion, "directory") + " in the storage: " + storageUrl);
+        }
+
+        if (requiredVersion != null) {
+            return new ExternalBinary(requiredVersion, storageUrl + matched.get(0).getKey());
+        } else {
+            final DriverEntry latestDriverEntry = findLatestDriver(matched);
+            return new ExternalBinary(latestDriverEntry.getLocation(), storageUrl + latestDriverEntry.getKey());
+        }
+    }
+
+    private DriverEntry findLatestDriver(List<DriverEntry> matched) {
+        return matched.stream().min((c1, c2) -> {
+            Date c1Date = parseDate(c1.getLastModified(), c1.getKey());
+            Date c2Date = parseDate(c2.getLastModified(), c2.getKey());
+            if (c1Date == null) return -1;
+            if (c2Date == null) return 1;
+            return Long.compare(c2Date.getTime(), c1Date.getTime());
+        }).get();
+    }
+
+    /**
+     * It is expected that this abstract method should return a regex that represents a key of an expected binary/file
+     * stored in the google storage. Key consist of <code>directory_name + / + file_name</code>. Keys are visible on
+     * the root of the google storage url (without index.html suffix)
+     *
+     * @param requiredVersion The required version set using method {@link XmlStorageSource#getReleaseForVersion},
+     *                        or otherwise null
+     * @param directory       The directory for the current binary the regex is being matched against
+     * @return A regex that represents a key of an expected binary/file stored in the google storage.
```

Comment:
The description still refers to google storage

Reply:
```suggestion
     * It is expected that this abstract method should return a regex that represents a key of an expected binary/file
     * stored in the xml storage. Commonly, key consists of <code>directory_name + / + file_name</code>. 
     *
     * @param requiredVersion The required version set using method {@link XmlStorageSource#getReleaseForVersion},
     *                        or otherwise null
     * @param directory       The directory for the current binary the regex is being matched against
     * @return A regex that represents a key of an expected binary/file.
```

Is it better?